src/Task/TalkNPC.pm

```diff
@@ -70,33 +80,107 @@ sub new {
+	$self->{wait_for_answer} = 0;
+	$self->{error_code} = undef;
+	$self->{error_message} = undef;
+	
 	my @holder = ($self);
 	Scalar::Util::weaken($holder[0]);
-	$self->{mapChangedHook} = Plugins::addHook('Network::Receive::map_changed', \&mapChanged, \@holder);
-
+	
+	push @{$self->{hookHandles}}, Plugins::addHooks(
+		['npc_talk',                  \&handle_npc_talk, \@holder],
+		['packet/npc_talk_continue',  \&handle_npc_talk, \@holder],
+		['npc_talk_done',             \&handle_npc_talk, \@holder],
+		['npc_talk_responses',        \&handle_npc_talk, \@holder],
+		['packet/npc_talk_number',    \&handle_npc_talk, \@holder],
+		['packet/npc_talk_text',      \&handle_npc_talk, \@holder],
+		['packet/npc_store_begin',    \&handle_npc_talk, \@holder],
+		['packet/npc_store_info',     \&handle_npc_talk, \@holder],
+		['packet/npc_sell_list',      \&handle_npc_talk, \@holder]
+	);
+	
 	return $self;
 }
 
+sub handle_npc_talk {
+	my ($hook_name, $args, $holder) = @_;
+	my $self = $holder->[0];
+	if ($hook_name eq 'npc_talk_done') {
+		$self->{stage} = AFTER_NPC_CLOSE;
+	}
+	$self->{time} = time;
+	$self->{sent_talk_response_cancel} = 0;
+	$self->{wait_for_answer} = 0;
+}
+
 sub DESTROY {
 	my ($self) = @_;
-	Plugins::delHook($self->{mapChangedHook}) if $self->{mapChangedHook};
+	debug "Destroying talknpc object\n", "ai_npcTalk";
+	Plugins::delHooks($_) for @{$self->{hookHandles}};
+	$self->SUPER::DESTROY;
 }
 
 # Overrided method.
 sub activate {
 	my ($self) = @_;
 	$self->SUPER::activate(); # Do not forget to call this!
 	$self->{time} = time;
-	$self->{stage} = 'Not Started';
-	$self->{mapChanged} = 0;
+	$self->{stage} = NOT_STARTED;
+}
+
+sub find_and_set_target {
+	my ($self) = @_;
+	my $target = $self->findTarget($npcsList);
+	if ($target) {
+		debug "Target NPC " . $target->name() . " at ($target->{pos}{x},$target->{pos}{y}) found.\n", "ai_npcTalk";
+	} else {
+		$target = $self->findTarget($monstersList);
+		if ($target) {
+			debug "Target Monster-NPC " . $target->name() . " at ($self->{pos}{x},$self->{pos}{y}) found.\n", "ai_npcTalk";
+		}
+	}
+
+	if ($target && $target->{statuses}->{EFFECTSTATE_BURROW}) {
+		$self->setError(NPC_NOT_FOUND, T("NPC is hidden."));
+		$target = undef;
+	}
+	
+	if (exists $talk{nameID}) {
+		if ($self->{type} eq 'talknpc') {
+			$self->{steps} = [parseArgs($self->{sequence})];
+		} else {
+			$self->{steps} = [];
+		}
+	} else {
+		if ($self->{type} eq 'talknpc') {
+			$self->{steps} = [parseArgs("x $self->{sequence}")];
+		} else {
+			push (@{$self->{steps}}, 'x');
+		}
+		undef $ai_v{'npc_talk'}{'time'};
+		undef $ai_v{'npc_talk'}{'talk'};
+	}
+	
+	if ($target) {
+		$self->{target} = $target;
+		$self->{ID} = $target->{ID};
+		lookAtPosition($self);
+	} elsif (exists $talk{nameID} && $ai_v{'npc_talk'}{'talk'} ne 'buy_or_sell') {#check if this is really necessary
+		$self->{ID} = $talk{ID};
+		$self->{target} = Actor::NPC->new;
+		$self->{target}->{appear_time} = time;
+		$self->{target}->{name} = 'Unknown';
```

Comment:
NPCs scripts that start conversation without an actual npc 'body', like automatic bot checkers and automatic npc conversations on map changes.